lib/invoiced.rb

```diff
@@ -95,7 +109,25 @@ def request(method, endpoint, params={})
             parse(response)
         end
 
-        private
+        def locked_api_request(method, endpoint, params)
+          lock = redis_lock
+          lock.lock
+            api_request(method, endpoint, params)
+          lock.unlock
+        rescue Redis::Lock::AcquireLockTimeOut
+          raise UnableToAcquireLock, "Acquire Lock timed out: #{lock.inspect}"
+        end
+
+        def redis_lock
+          return unless @redis
+
+          Redis::Lock.new(
+            @redis,
+            "invoiced_lock_#{rand(1..ConnectionLimit)}",
```

Comment:
Related to that configuration, perhaps a lock namespace or some other value to prevent accidental collisions with other locks. There might be an invoiced identifier we can use for that (e.g. an account id or something to separate locking across separate limits)